app/models/dictionary_sign_asset.rb

```diff
@@ -0,0 +1,57 @@
+class DictionarySignAsset
+  attr_reader :asset_url
+
+  class S3Adapter
+    cattr_accessor :region, :access_key_id, :secret_access_key
+    self.region = ENV.fetch("DICTIONARY_AWS_REGION", ENV.fetch("AWS_REGION", nil))
+    self.access_key_id = ENV.fetch("DICTIONARY_AWS_ACCESS_KEY_ID", nil)
+    self.secret_access_key = ENV.fetch("DICTIONARY_AWS_SECRET_ACCESS_KEY", nil)
+
+    def initialize(asset)
+      @asset = asset
+    end
+
+    def self.configured?
+      client.present?
+    end
+
+    def self.client
+      @client ||= Aws::S3::Client.new(region:, access_key_id:,
+                                      secret_access_key:)
+    rescue Aws::Sigv4::Errors::MissingCredentialsError, Aws::Sigv4::Errors::MissingRegionError
+      nil
+    end
+
+    def url(expires_in: 1.hour)
+      return unless self.class.configured?
+
+      bucket_name = @asset.asset_url.host.split(".")[0]
```

Comment:
This extraction assumes a particular URL shape. I didn't add any assertion type stuff here, but I do wonder if it would be safer to raise an ArgumentError if the URL was not in the form I expect it to be. I'd prefer to do this by asserting on the hostname, since a regexp-based solution could get quite complex.

Reply:
Implemented a hostname check in 8cf962d4